### PR TITLE
BLS precompile native benchmarking & improvements

### LINKED ISSUE #7529: Benchmarks and gas costs of 2537

### PR DESCRIPTION
**Pull request by Marchhill**
Resolves #7529

## Changes

- Benchmark BLS precompile with BenchmarkDotNet ([results](https://github.com/Marchhill/bls-precompile-benchmarks/tree/main))
- Rename precompiles and tests for consistency with EIP
- Add options to disable concurrency and subgroup checks
- Optimisations and refactoring
- Move vectors to [bench_precompiles](https://github.com/NethermindEth/bench_precompiles)

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [ ] Yes
- [x] No

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by MarekM25 — APPROVED** [review]
This PR can be merged in this form, you added metrics which is good

**Comment by Marchhill** [comment]
> This PR can be merged in this form, you added metrics which is good

Moved the metrics to a different PR: https://github.com/NethermindEth/nethermind/pull/7546

**Review by LukaszRozmej — APPROVED** [review]

**Review by MarekM25 — COMMENTED** [review]
Do you have results? :)

File: src/Nethermind/Nethermind.Evm/Precompiles/Bls/G1MSMPrecompile.cs [hunk 1]
```diff
@@ -96,4 +99,28 @@ public long DataGasCost(ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec
         G1 res = new G1().MultiMult(rawPoints.AsSpan(), rawScalars.AsSpan(), npoints);
         return (res.EncodeRaw(), true);
     }
+
+    private static bool TryDecodePoint(ReadOnlyMemory<byte> inputData, Span<long> rawPoints, Span<byte> rawScalars, int dest, int i)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Precompiles/Bls/G1MSMPrecompile.cs:104]: [SkipLocalsInit]

File: src/Nethermind/Nethermind.Evm/Precompiles/Bls/G2MSMPrecompile.cs [hunk 2]
```diff
@@ -97,4 +100,28 @@ public long DataGasCost(ReadOnlyMemory<byte> inputData, IReleaseSpec releaseSpec
 
         return (res.EncodeRaw(), true);
     }
+
+    private static bool TryDecodePoint(ReadOnlyMemory<byte> inputData, Span<long> rawPoints, Span<byte> rawScalars, int dest, int i)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/Precompiles/Bls/G2MSMPrecompile.cs:105]: Almost duplicate code, maybe can be done with generics?

**Review by LukaszRozmej — COMMENTED** [review]
should we bring those test files directly, are they available in some repo that we could bring as a submodule?

**Comment by Marchhill** [comment]
> should we bring those test files directly, are they available in some repo that we could bring as a submodule?

they are in benchmark_precompiles, I could either add the new bls vectors there or copy them all to the same directory as I did

File: src/Nethermind/Nethermind.Evm/Precompiles/Bls/G2MSMPrecompile.cs (hunk 2 above)
Comment (Marchhill) [reply src/Nethermind/Nethermind.Evm/Precompiles/Bls/G2MSMPrecompile.cs]: Did it by passing a function instead

File: src/Nethermind/Nethermind.Evm/Precompiles/Bls/G1MSMPrecompile.cs (hunk 1 above)
Comment (Marchhill) [reply src/Nethermind/Nethermind.Evm/Precompiles/Bls/G1MSMPrecompile.cs]: Added elsewhere

**Pull request merged**